### PR TITLE
AbsSightline

### PR DESCRIPTION
**Pull request by profxj**
Creates AbsSightline class.
It is mainly a simple object to hold a set of 
AbsComponent objects along a sightline on the sky.

A few other updates snuck in as well.

Tests and docs.

### CODE REVIEW

File: docs/abssightline.rst [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+	abssl = GenericAbsSightline((10.0*u.deg, 45*u.deg))
+
+More commonly, one will instantiate with one
+a set of components::
+
+    lya = AbsLine('HI 1215')
+    lya.attrib['z'] = 2.3
+    lyb = AbsLine(1025.7222*u.AA, z=lya.attrib['z'])
+    lyb.limits.set([-300.,300.]*u.km/u.s)  # vlim
+    abscomp = AbsComponent.from_abslines([lya,lyb])
```
Comment (jnburchett) [inline docs/abssightline.rst]: If I do not also set lya.limits, I get the following warning and the resulting abscomp has [0,0] velocity limits:

Input absline with wrest=1025.72 Angstrom at z=2.300 does not match component rules. Not appending
Absline velocities lie beyond component
 Set chk_vel=False to skip this test.

File: docs/examples/AbsSightline.ipynb [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+      ]
+     },
+     "execution_count": 14,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "reload(lt_abss)\n",
+    "gensl = lt_abss.GenericSightline('J140923.897+261821.043')\n",
```
Comment (jnburchett) [inline docs/examples/AbsSightline.ipynb]: Should be 'lt_abss.GenericAbsSightline'....

File: linetools/isgm/abssightline.py [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+    Ej : Quantity
+        Energy of lower level (1/cm)
+    comment : str
+        A comment, default is ``
+    """
+    __metaclass__ = ABCMeta
+
+    @classmethod
+    def from_abslines(cls, abslines, stars=None, **kwargs):
+        """Instantiate from a list of AbsLine objects
```
Comment (ntejos) [inline linetools/isgm/abssightline.py]: This code is duplicated from elsewhere, we may want to move these to utils and call it from there for `AbsSystem` and `AbsComponent` and `AbsSightline`.

File: linetools/isgm/abssightline.py [hunk 4]
```diff
@@ -0,0 +1,242 @@
...
+        comps = build_components_from_abslines(abslines)
+
+        slf = cls.from_components(comps)
+
+        # Return
+        return slf
+
+    @classmethod
+    def from_components(cls, components, **kwargs):
+        """ Instantiate from a list of AbsComponent objects
```
Comment (ntejos) [inline linetools/isgm/abssightline.py]: Ditto.

File: linetools/isgm/abssightline.py [hunk 5]
```diff
@@ -0,0 +1,242 @@
...
+        """  Initiator
+
+        Parameters
+        ----------
+        radec : tuple or SkyCoord
+            (RA,DEC) in deg or astropy.coordinate
+        sl_type : str, optional
+          Sightline type, e.g. IGM
+        emtype : str, optional
+          Type of emission source for absorption sightline
```
Comment (ntejos) [inline linetools/isgm/abssightline.py]: e.g. 'QSO'

File: linetools/isgm/abssightline.py [hunk 6]
```diff
@@ -0,0 +1,242 @@
...
+    def __init__(self, radec, sl_type=None, em_type=None, comment=None, name=None):
+        """  Initiator
+
+        Parameters
+        ----------
+        radec : tuple or SkyCoord
+            (RA,DEC) in deg or astropy.coordinate
+        sl_type : str, optional
+          Sightline type, e.g. IGM
+        emtype : str, optional
```
Comment (ntejos) [inline linetools/isgm/abssightline.py]: `em_type`

File: linetools/isgm/abssightline.py [hunk 7]
```diff
@@ -0,0 +1,242 @@
...
+        # Return
+        return slf
+
+    def __init__(self, radec, sl_type=None, em_type=None, comment=None, name=None):
+        """  Initiator
+
+        Parameters
+        ----------
+        radec : tuple or SkyCoord
+            (RA,DEC) in deg or astropy.coordinate
```
Comment (ntejos) [inline linetools/isgm/abssightline.py]: astropy.coordinate.SkyCoord

File: linetools/isgm/abssightline.py [hunk 8]
```diff
@@ -0,0 +1,242 @@
...
+        emtype : str, optional
+          Type of emission source for absorption sightline
+        comment : str, optional
+          A comment, default is ``
+        """
+        # Required
+        self.coord = ltu.radec_to_coord(radec)
+
+        # Lists
+        self._components = []
```
Comment (ntejos) [inline linetools/isgm/abssightline.py]: We may want to consider to have also AbsSystems within the AbsSightline

File: linetools/isgm/abssightline.py [hunk 9]
```diff
@@ -0,0 +1,242 @@
...
+
+        # Name
+        if name is None:
+            self.name = 'J{:s}{:s}'.format(
+                    self.coord.ra.to_string(unit=u.hour,sep='',pad=True),
+                    self.coord.dec.to_string(sep='',pad=True,alwayssign=True))
```
Comment (ntejos) [inline linetools/isgm/abssightline.py]: Consider moving this naming code to a method called from linetools.utils, e.g.  `ltu.coord_to_str()` or similar.

File: linetools/isgm/abssightline.py [hunk 10]
```diff
@@ -0,0 +1,242 @@
...
+        else:
+            self.name=name
```
Comment (ntejos) [inline linetools/isgm/abssightline.py]: Check whether name is str (?)

File: linetools/isgm/abssightline.py [hunk 11]
```diff
@@ -0,0 +1,242 @@
...
+
+        # Others
+        self.em_type = em_type
+        self.sl_type = sl_type
```
Comment (ntejos) [inline linetools/isgm/abssightline.py]: Ditto.

File: linetools/isgm/abssightline.py [hunk 12]
```diff
@@ -0,0 +1,242 @@
...
+
+    def add_component(self, abscomp, tol=0.2*u.arcsec,
+                      chk_sep=True, debug=False, **kwargs):
+        """Add an AbsLine object to the component if it satisfies
```
Comment (ntejos) [inline linetools/isgm/abssightline.py]: AbsLine -> AbsComponent

Comment (ntejos) [inline linetools/isgm/abssightline.py]: component -> AbsSightline

File: linetools/isgm/abssightline.py [hunk 13]
```diff
@@ -0,0 +1,242 @@
...
+        all of the rules.
+
+        For velocities, we demand that the new line has a velocity
```
Comment (ntejos) [inline linetools/isgm/abssightline.py]: Remove this paragraph.

File: linetools/isgm/abssightline.py [hunk 14]
```diff
@@ -0,0 +1,242 @@
...
+        range that is fully encompassed by the component.
+
+        Parameters
+        ----------
+        abscomp : AbsComp
```
Comment (ntejos) [inline linetools/isgm/abssightline.py]: AbsComponent

File: linetools/isgm/abssightline.py [hunk 15]
```diff
@@ -0,0 +1,242 @@
...
+        chk_sep : bool, optional
+          Perform coordinate check (expensive)
+        """
+        # Coordinates
+        if chk_sep:
+            testcoord = bool(self.coord.separation(abscomp.coord) < tol)
+        else:
+            testcoord = True
+
+        # Combine
```
Comment (ntejos) [inline linetools/isgm/abssightline.py]: Combine with potential other tests

File: linetools/isgm/abssightline.py [hunk 16]
```diff
@@ -0,0 +1,242 @@
...
+        test = testcoord
+        # Append?
+        if test:
+            self._components.append(abscomp)
+        else:
+            warnings.warn("Failed add_component test")
+            print('Input absline with wrest={:g} does not match component rules. Not appending'.format(absline.wrest))
```
Comment (ntejos) [inline linetools/isgm/abssightline.py]: absline -> AbsComponent

File: linetools/isgm/abssightline.py [hunk 17]
```diff
@@ -0,0 +1,242 @@
...
+            if not testcoord:
+                print("AbsComp coordinates do not match.  Best to set them")
+
+    def build_table(self):
```
Comment (ntejos) [inline linetools/isgm/abssightline.py]: I think there should be a generic function in isgm.utils that receives a list of components and creates the table. This method could call that function.

Comment (ntejos) [inline linetools/isgm/abssightline.py]: Consider a name like: `table_from_abscomponents()`

File: linetools/isgm/abssightline.py [hunk 18]
```diff
@@ -0,0 +1,242 @@
...
+        for component in self._components:
+            outdict['components'][component.name] = ltu.jsonify(component.to_dict())
+        # Polish
+        outdict = ltu.jsonify(outdict)
+        # Return
+        return outdict
+
+    def __repr__(self):
+        txt = '<{:s}: {:s} {:s}'.format(
+            self.__class__.__name__, self.coord.ra.to_string(unit=u.hour,sep=':', pad=True),
```
Comment (ntejos) [inline linetools/isgm/abssightline.py]: Shouldn't the representation use the name given by the user? Maybe the solution is to have two names: (i) linetools name JXXXXXX.XX+XXXXXX and the user defined name. The representation within linetools is always with the lt preferred format.

File: linetools/isgm/abssightline.py [hunk 19]
```diff
@@ -0,0 +1,242 @@
...
+        if self.em_type is not None:
+            txt = txt + ', emtype={:s}'.format(self.em_type)
+
+        # Finish
+        txt += '>'
+        return (txt)
+
+
+class GenericAbsSightline(AbsSightline):
+    """Class for Generic Absorption Sightline
```
Comment (ntejos) [inline linetools/isgm/abssightline.py]: What is the idea of this Class? I think it should be a GenericSightline class, and then AbsSightline would be a child class of the generic one. But you may have another idea in mind...

File: linetools/isgm/tests/test_init_abssightline.py [hunk 20]
```diff
@@ -0,0 +1,86 @@
...
+        iline = AbsLine(trans)
+        iline.attrib['z'] = 2.92939
+        iline.attrib['coord'] = radec
+        iline.analy['vlim'] = [-250.,80.]*u.km/u.s
+        abslines.append(iline)
+    #
+    SiII_comp = AbsComponent.from_abslines(abslines)
+    #
+    return SiII_comp
+
```
Comment (ntejos) [inline linetools/isgm/tests/test_init_abssightline.py]: Are the above tests duplication from elsewhere? If so, please delete or move to the corresponding place.

File: linetools/isgm/tests/test_use_abssightline.py [hunk 21]
```diff
@@ -0,0 +1,81 @@
...
+        iline.analy['vlim'] = [-250.,80.]*u.km/u.s
+        abslines.append(iline)
+    #
+    SiII_comp = AbsComponent.from_abslines(abslines)
+    SiII_comp.logN = 15.
+    SiII_comp.flag_N = 1
+    #
+    return SiII_comp
+
+
```
Comment (ntejos) [inline linetools/isgm/tests/test_use_abssightline.py]: Ditto.

File: docs/examples/AbsSightline.ipynb (hunk 2 above)
Comment (profxj) [inline docs/examples/AbsSightline.ipynb]: Right, fixed

File: docs/abssightline.rst (hunk 1 above)
Comment (profxj) [inline docs/abssightline.rst]: There were several issues here.  Fixed, thanks.

File: linetools/isgm/abssightline.py [hunk 22]
```diff
@@ -0,0 +1,242 @@
+""" Classes for absorption line component
+"""
```
Comment (profxj) [inline linetools/isgm/abssightline.py]: fixed.

File: linetools/isgm/abssightline.py [hunk 23]
```diff
@@ -0,0 +1,242 @@
...
+        e.g. (8,1) for OI
+    zcomp : float
+        Component redshift
+    vlim : Quantity array
+        Velocity limits of the component
+        e.g.  [-300,300]*u.km/u.s
+    A : int
+        Atomic mass -- used to distinguish isotopes
+    Ej : Quantity
+        Energy of lower level (1/cm)
```
Comment (profxj) [inline linetools/isgm/abssightline.py]: All of these fixed.  Sloppy on my part

File: linetools/isgm/abssightline.py (hunk 3 above)
Comment (profxj) [inline linetools/isgm/abssightline.py]: Actually, little is duplicated in this method.  Only error checking.

File: linetools/isgm/abssightline.py (hunk 4 above)
Comment (profxj) [inline linetools/isgm/abssightline.py]: I don't think we can put a classmethod in utils.

File: linetools/isgm/abssightline.py (hunk 7 above)
Comment (profxj) [inline linetools/isgm/abssightline.py]: fixed

File: linetools/isgm/abssightline.py (hunk 5 above)
Comment (profxj) [inline linetools/isgm/abssightline.py]: added

File: linetools/isgm/abssightline.py (hunk 8 above)
Comment (profxj) [inline linetools/isgm/abssightline.py]: Agreed, and I almost added it.  You've
convinced me to set it to None just to save the 
namespace.

File: linetools/isgm/abssightline.py (hunk 9 above)
Comment (profxj) [inline linetools/isgm/abssightline.py]: good suggestion; long overdue.  added  
name_from_coord

File: linetools/isgm/abssightline.py (hunk 10 above)
Comment (profxj) [inline linetools/isgm/abssightline.py]: But these are all default to None

File: linetools/isgm/abssightline.py (hunk 12 above)
Comment (profxj) [inline linetools/isgm/abssightline.py]: fixed

File: linetools/isgm/abssightline.py (hunk 14 above)
Comment (profxj) [inline linetools/isgm/abssightline.py]: fixed

File: linetools/isgm/abssightline.py (hunk 15 above)
Comment (profxj) [inline linetools/isgm/abssightline.py]: edited

File: linetools/isgm/abssightline.py (hunk 16 above)
Comment (profxj) [inline linetools/isgm/abssightline.py]: commented

File: linetools/isgm/abssightline.py (hunk 18 above)
Comment (profxj) [inline linetools/isgm/abssightline.py]: That is only the class name (e.g. GenericAbsSightline).

I am not providing the name here.  Only the RA+DEC.
 But now I am.  ;)

File: linetools/isgm/abssightline.py (hunk 19 above)
Comment (profxj) [inline linetools/isgm/abssightline.py]: There are only AbsSighltines, not EmSightlines.

So the Generic one is GenericAbsSightline

File: linetools/isgm/tests/test_init_abssightline.py (hunk 20 above)
Comment (profxj) [inline linetools/isgm/tests/test_init_abssightline.py]: The above are defs, not tests.  Have pushed to a utils.py
module in the tests folder

**Comment by profxj** [comment]
Lots of edits addressing NT and JB comments. 

Thanks for those!

File: linetools/isgm/abssightline.py [hunk 24]
```diff
@@ -110,11 +104,11 @@ def __init__(self, radec, sl_type=None, em_type=None, comment=None, name=None):
...
         radec : tuple or SkyCoord
-            (RA,DEC) in deg or astropy.coordinate
+            (RA,DEC) in deg or astropy.coordinate.SkyCoord
         sl_type : str, optional
           Sightline type, e.g. IGM
-        emtype : str, optional
-          Type of emission source for absorption sightline
+        em_type : str, optional
+          Type of emission source for absorption sightline, e.g. QSO
         comment : str, optional
```
Comment (ntejos) [inline linetools/isgm/abssightline.py]: `name` missing in docstring.

File: linetools/isgm/abssightline.py [hunk 25]
```diff
@@ -126,27 +120,24 @@ def __init__(self, radec, sl_type=None, em_type=None, comment=None, name=None):
...
-                    self.coord.dec.to_string(sep='',pad=True,alwayssign=True))
+            self.name = ltu.name_from_coord(self.coord)
         else:
-            self.name=name
+            self.name = name
 
         # Others
         self.em_type = em_type
         self.sl_type = sl_type
+        self._abssystems = None
```
Comment (ntejos) [inline linetools/isgm/abssightline.py]: add comment like: potentially we may want to save AbsSystems in this class but not implemented for now.

File: linetools/isgm/abssightline.py [hunk 26]
```diff
@@ -126,27 +120,24 @@ def __init__(self, radec, sl_type=None, em_type=None, comment=None, name=None):
...
 
     def add_component(self, abscomp, tol=0.2*u.arcsec,
                       chk_sep=True, debug=False, **kwargs):
-        """Add an AbsLine object to the component if it satisfies
-        all of the rules.
+        """Add a component to AbsSightline if it satisfies all of the rules.
 
-        For velocities, we demand that the new line has a velocity
-        range that is fully encompassed by the component.
+        Presently, the only constraint is on RA/DEC
```
Comment (ntejos) [inline linetools/isgm/abssightline.py]: Now that I think about it, we may also want to impose that the redshift of the AbsComponent is within certain tolerance from z_source or lower. We may need to store z_source as an attribute of the Class.

File: linetools/isgm/abssightline.py [hunk 27]
```diff
@@ -225,6 +210,9 @@ def __repr__(self):
             self.__class__.__name__, self.coord.ra.to_string(unit=u.hour,sep=':', pad=True),
                 self.coord.dec.to_string(sep=':',pad=True,alwayssign=True))
 
+        # Name
+        if self.name is not None:
+            txt = txt + ', name={:s}'.format(self.name)
```
Comment (ntejos) [inline linetools/isgm/abssightline.py]: Not sure if `self.name` can be `None` in the current implementation.

File: linetools/isgm/abssightline.py (hunk 24 above)
Comment (profxj) [inline linetools/isgm/abssightline.py]: added.

File: linetools/isgm/abssightline.py (hunk 25 above)
Comment (profxj) [inline linetools/isgm/abssightline.py]: added

File: linetools/isgm/abssightline.py (hunk 26 above)
Comment (profxj) [inline linetools/isgm/abssightline.py]: I thought about this and decided that the source definition may vary from sub-class, e.g.
z for IGM sightlines, v or d for ISM sightlines.  Therefore decided to let the sub-class 
deal with this.

Comment (profxj) [inline linetools/isgm/abssightline.py]: Added a comment.

File: linetools/isgm/abssightline.py (hunk 27 above)
Comment (profxj) [inline linetools/isgm/abssightline.py]: It is defaulted to None.

**Comment by ntejos** [comment]
Sounds good.

--
Nicolas Tejos, Ph.D.
Postdoctoral Researcher
Millennium Institute of Astrophysics
Santiago, Chile

File: linetools/isgm/abssightline.py (hunk 8 above)
Comment (jnburchett) [inline linetools/isgm/abssightline.py]: Yes, this is a good suggestion

File: docs/abssightline.rst [hunk 28]
```diff
@@ -0,0 +1,75 @@
...
+The default sets the properties listed above::
+
+	abssl = GenericAbsSightline((10.0*u.deg, 45*u.deg))
+
+More commonly, one will instantiate with one
+a set of components::
+
+   lya = AbsLine('HI 1215')
+    lya.limits.set([-300.,300.]*u.km/u.s)  # vlim
+    lya.attrib['z'] = 2.3
```
Comment (jnburchett) [inline docs/abssightline.rst]: When I check `lya.limits`, I still get weird zlim values:
```
In [56]: lya.limits
Out[56]: <LineLimits wrest=1215.67 Angstrom z=0 zlim=[-0.00100019  0.00100119] wvlim=[ 1214.45409648  1216.88712088] Angstrom vlim=[-300.  300.] km / s>
```

Comment (profxj) [inline docs/abssightline.rst]: Yeah, this is an issue but not with AbsSightline.  Am going to deal with it in a separate PR

Comment (profxj) [inline docs/abssightline.rst]: But also modify the Notebook here accordingly.

Comment (jnburchett) [inline docs/abssightline.rst]: Good point!

**Comment by profxj** [comment]
Merging

**Pull request merged**